lua/pqf/init.lua

```diff
@@ -88,7 +93,26 @@ local function pad_right(string, pad_to)
 end
 
 local function trim_path(path)
-  return fn.fnamemodify(path, ':p:.')
+  local fname = fn.fnamemodify(path, ':p:.')
+  local len = fn.strchars(fname)
+  local truncate_to = max_filename_length
+  if max_filename_length > 0 and max_filename_length < 1 then -- percentage
+    truncate_to = math.floor(max_filename_length * fn.winwidth(0))
+  end
+  if truncate_to > 0 and len > truncate_to then
+    if filename_truncate_position == 'start' then
+      fname = '...'
+        .. fn.strpart(fname, len - truncate_to, truncate_to, vim.v['true'])
+    elseif filename_truncate_position == 'end' then
+      fname = fn.strpart(fname, 0, truncate_to, vim.v['true']) .. '...'
+    elseif filename_truncate_position == 'middle' then
+      local halflen = math.floor(truncate_to / 2)
+      fname = fn.strpart(fname, 0, halflen, vim.v['true'])
+        .. '...'
+        .. fn.strpart(fname, len - halflen, halflen, vim.v['true'])
+    end
+  end
+  return fname
```

Comment:
This is way too much logic for my liking, and combined with the lack of empty lines between e.g. the `if` expressions, it's really hard to read.

Setting that aside, using percentage won't work reliably, because when resizing the window the limits aren't resized either (nor would I want to, because it ends up being far too complex to implement).

I would like for a more opinionated and less complex approach: lines are limited to fixed amount of characters/columns or none at all, but no percentage or anything like that. The ellipsis should be at the start and only at the start. Putting them at the end would end up hiding the file name, which is never what you want. Similarly, putting them in the middle doesn't strike me as very useful, as in a very long path the first few components shown probably aren't helpful (and likely the same as many other entries). 

Finally, I think `...` could be mistaken for being part of the file path itself, as it's too similar to `..`. I think using `[...]` would make it more obvious.

Thus, you'd end up with something like this:

```
[...]src/type_check/expressions.rs:2221:66          for (patt, member) in node.values.iter_mut().zip(members.into_iter()) {
[...]src/type_check/expressions.rs:2476:22          // moved into the closure may still be referred to from elsewhere.
```

I realize this may not be quite what you had in mind or need, but I want to keep my plugins (including this one) as simple and easy to maintain as possible, and the current approach isn't close enough to that for my liking :)